Feature/shop

pubestpubest opened the pull request:
# Pull Request: Shop Management Functionality  

This PR adds shop and product management features, including creation, updates, deletion, and authentication.  

## Key Changes  

- **Shop Management:** Introduced `ShopUsecase` and `ShopRepository` interfaces (`domain/shop.go`).  
- **Entity Updates:** Updated `Shop` struct (`entity/shop.go`) with `Password` and constraints.  
- **HTTP Handlers:** Implemented endpoints for shop and product management (`features/shop/delivery/http.go`).  
- **Repository & Use Case:** Added `shopRepository` (`repository/postgres.go`) and `shopUsecase` (`usecase/usecase.go`) for database operations and business logic.  
- **JWT Authentication:** Integrated JWT for shop authentication.  
- **Standardized Responses:** Defined API response structs for `Product` and `Shop`.  
- **Dependencies & Integration:** Added `github.com/golang-jwt/jwt/v5` and integrated features into `main.go`.

pubestpubest committed:
feat: added
- get all shop [circular preload needed to be fix]
- create shop
- create product

pubestpubest committed:
fix: instead of using preload, we build it from separate repos

pubestpubest committed:
feat: enhance shop registration with password requirement and unique name constraint

pubestpubest committed:
feat: add login functionality and shop retrieval by name, along with product fetching by shop ID

pubestpubest committed:
feat: implement product update functionality and add product retrieval by shop ID
- shops JWT extractor needed to be fix

pubestpubest committed:
chore: update .gitignore to include .env and add golang-jwt/v5 dependency in go.mod and go.sum

pubestpubest committed:
fix: shops JWT extractor using v5

pubestpubest committed:
refactor: no longer use claims to held products
- update shop usecase and repository to support product retrieval by ID and enhance shop response structure with products

pubestpubest committed:
feat: add product deletion functionality and enhance shop profile retrieval
- implemented DeleteProduct method in usecase and repository
- updated HTTP handler to support product deletion and shop profile retrieval with JWT validation

pubestpubest committed:
refac: update shop retrieval to include products and enhance HTTP routes
- modified GetShopByName to return ShopWithProducts in usecase and repository
- updated HTTP routes to reflect new product management structure under /shops/products
- removed unused GetProducts handler from HTTP delivery

pubestpubest committed:
refactor: restructure HTTP routes for shops to separate public and authenticated access
- created public and authenticated groups for shop-related endpoints
- replaced direct JWT claims retrieval with middleware for authorization checks
- removed unused readToken function to streamline authentication process

tananut-pdkm commented on features/shop/delivery/http.go:
```diff
@@ -0,0 +1,215 @@
+		"description": shop.Description,
+		"products":    products,
+	}
+	token := jwt.NewWithClaims(jwt.SigningMethodHS256, jwt.MapClaims{
```
We need to move the claim logic into usecase session instead of delivery and we can group the claim JWT token into utils function [optional].

pubestpubest replied:
This should resolved by 17b75e0de5b8239c33a37692891b8e23a1a34a4d

tananut-pdkm commented on main.go:
```diff
@@ -115,6 +118,12 @@ func main() {
 		),
 	)
 
+	shopDelivery.NewHandler(v1,
```
We need to create to route group that name `shopGroup` to split the party routes and use `shopGroup` instead of `v1` route

pubestpubest replied:
This should resolved by ffceb7c0acb04ed7e685caa378c02be162d5cb1f

tananut-pdkm commented on features/shop/delivery/http.go:
```diff
@@ -0,0 +1,215 @@
+}
+
+func (h *Handler) GetShopProfile(c echo.Context) error {
+	claims, ok := c.Get("shop").(*response.Shop)
```
We should rename from `claims` -> `shop`

pubestpubest replied:
This should resolved by fe40136fc737f0b9f96d0115f668263f0a79c45c

tananut-pdkm commented on domain/shop.go:
```diff
@@ -0,0 +1,31 @@
+	GetShopByName(name string) (response.ShopWithProducts, error)
+	Login(name string, password string) (entity.Shop, error)
+	GetProductsByShopID(id uint32) ([]response.Product, error)
+	BelongsToShop(productID uint32, claims *response.Shop) bool
```
We don't need this function and we should add the Where clause query with shop instead for example `DELETE FROM products WHERE id = ? AND shop_id = ?`

pubestpubest replied:
This commit should resolve this 4d6f80a23fa9abed1df2177eee9e4ed777909027 😃

tananut-pdkm commented:
- We should remove the cookie system
- We have to refactor the authentication system

tananut-pdkm commented on features/shop/delivery/http.go:
```diff
@@ -0,0 +1,215 @@
+	if err != nil {
+		return c.JSON(http.StatusInternalServerError, err.Error())
+	}
+	if err := bcrypt.CompareHashAndPassword([]byte(shop.Password), []byte(req.Password)); err != nil {
```
We need to move the password comparing logic into usecase session

pubestpubest replied:
This should resolved by 17b75e0de5b8239c33a37692891b8e23a1a34a4d

pubestpubest committed:
refac: group reponse to entity

pubestpubest committed:
refac: refactored repo query and deleted belongto

pubestpubest committed:
refactor: update shop login process to use JWT and streamline authentication
- modified Login method to return JWT token instead of shop entity
- adjusted HTTP handler to set Authorization header with token
- implemented utility functions for JWT generation and validation
- removed unused middleware and streamlined route handling for authentication

pubestpubest committed:
refactor: reorganized group route

pubestpubest committed:
refactor: improve variable naming and code readability in shop delivery and usecase layers

pubestpubest committed:
feat: enhance error handling and response structure

- added error wrapping for better traceability in repository methods
- updated HTTP handlers to return structured responses with success status and messages
- introduced ShopExists method in repository to check shop existence before operations

pubestpubest commented:
### 🔁 PR Update Summary

This PR has been updated and is ready for re-review. Here's a summary of the recent changes:

---

#### ✅ Refactors
- **Reorganized group routes** for better structure and modularity. (`17b75e0`)
- **Improved variable naming and code readability** in `shop/delivery`. (`ffceb7c`)
- **Refactored repository queries** and **removed unused `belongTo` logic**. (`bb725ac`)
- **Updated shop login process** to use **JWT-based authentication**, streamlining the auth flow. (`4d6f80a`)

#### ✨ Features
- Enhanced **error handling** and **response structure** for consistency and clarity across handlers. (`fe40136`)

---

pubestpubest committed:
refactor: rename ShopResponse to ShopWithOutPassword

pubestpubest committed:
refac: enhance utils/jwt for both party, delete redundant product.ID and shop.ID, refactor the product management struct to only id

pubestpubest committed:
fix: parse jwt.mapClaims to ShopWithOutPassword

pubestpubest committed:
fix: archived auth.go and add buffer to quit channel for graceful shutdown

pubestpubest committed:
Merge branch 'develop' into feature/shop
